johnzon-mapper/src/main/java/org/apache/johnzon/mapper/MappingParserImpl.java

```diff
@@ -399,36 +403,43 @@ private Object buildObject(final Type inType, final JsonObject object, final boo
-                value.writer.write(t, jsonValue);
-                continue;
-            }
-            if (jsonValue == null) {
-                continue;
-            }
+            try {
+                if (JsonValue.class == value.paramType) {
+                    value.writer.write(t, jsonValue);
+                    continue;
+                }
+                if (jsonValue == null) {
+                    continue;
+                }
 
-            final AccessMode.Writer setterMethod = value.writer;
-            if (NULL == valueType) { // forced
-                setterMethod.write(t, null);
-            } else {
-                Object existingInstance = null;
-                if (config.isReadAttributeBeforeWrite()) {
-                    final Mappings.Getter getter = classMapping.getters.get(jsonEntry.getKey());
-                    if (getter != null) {
-                        try {
-                            existingInstance = getter.reader.read(t);
-                        } catch (final RuntimeException re) {
-                            // backward compatibility
+                final AccessMode.Writer setterMethod = value.writer;
+                if (NULL == valueType) { // forced
+                    setterMethod.write(t, null);
+                } else {
+                    Object existingInstance = null;
+                    if (config.isReadAttributeBeforeWrite()) {
+                        final Mappings.Getter getter = classMapping.getters.get(jsonEntry.getKey());
+                        if (getter != null) {
+                            try {
+                                existingInstance = getter.reader.read(t);
+                            } catch (final RuntimeException re) {
+                                // backward compatibility
+                            }
                         }
                     }
+                    final Object convertedValue = toValue(
+                            existingInstance, jsonValue, value.converter, value.itemConverter,
+                            value.paramType, value.objectConverter,
+                            isDedup() ? new JsonPointerTracker(jsonPointer, jsonEntry.getKey()) : null, inType);
+                    if (convertedValue != null) {
+                        setterMethod.write(t, convertedValue);
+                    }
                 }
-                final Object convertedValue = toValue(
-                        existingInstance, jsonValue, value.converter, value.itemConverter,
-                        value.paramType, value.objectConverter,
-                        isDedup() ? new JsonPointerTracker(jsonPointer, jsonEntry.getKey()) : null, inType);
-                if (convertedValue != null) {
-                    setterMethod.write(t, convertedValue);
-                }
+            } catch (SetterMappingException alreadyHandled) {
+                throw alreadyHandled;
+            } catch (Exception e) {
```

Comment:
Shouldnt it let most RuntimeException go to not break exception handling in user code?

Reply:
I know internally we throw a lot of RuntimeExceptions like IllegalArgumentException and the NullPointerException I fixed still got wrapped.  To be extra sure I just updated #90 with a new test case that checks how we handle it when users throw a RuntimeException from their setters.  Looks like we always wrap their RuntimeException with JsonbException and generate a new message.

Reply:
I know but in mapper we let runtime go.
More than happy if we subclass java.lang.XException to catch the cases you mentionned but user exceptions (converter, adapter, ...) nor johnzon exception unrelated to that handling should be wrapped IMHO.

Reply:
Ok, to make sure I understand correctly, given code like the following (from the website):

```
final Mapper mapper = new MapperBuilder().build();
final MySuperObject otherObject = mapper.readObject(inputStream, MySuperObject.class);
```
If MySuperObject throws a RuntimeException from its setter or constructor that exact exception instance can be caught if `readObject` was wrapped in a try-catch.  Is that basically right?

Reply:
I created a test in PR #90 in the `johnzon-mapper` module that doesn't use JSON-B and instead uses `Mapper.readObject`.  The handling seems identical and any RuntimeException thrown from the user's code is wrapped in a `MappingException`.  If there is still a scenario I'm missing, the most helpful way forward would be if you hack up a small test case I can use as a reference.

Reply:
@dblevins so you mean you didn't change the exception stacktrace at all? If so fine for me (indeed it includes `getCause()` ;)). It is indeed about setters+getters (`write`/`read`) but also `Adapter`/`Converter`s (`toValue`). Think we should limit our breaking change to subclass existing exception in our own codebase but not force users to change their specific exception handling. A quick option is to refine the try/catch block to only needed parts I guess and passthrough `RuntimeException` as this which should enable to handle the cases you want to catch without side effects and other changes.

Reply:
A compassionate request to not use ";)" while we work this out as I find the explanations unclear and it makes me feel teased.

On the code I dig deeper as to how a user can find their own RuntimeException on a `mapper.readObject` call:

 - Thrown from a Converter: `catch (MapperException e) { e.getCause(); }` will return the RuntimeException
 - Thrown from a setter: `catch (MapperException e) { e..getCause().getCause(); }` will return the RuntimeException
 - Thrown from a constructor: `catch (IllegalStateException e) { e.getCause(); }` will return the RuntimeException

With all the scenarios handled very differently, I'm not sure I see a clear path forward that doesn't involve some kind of change.

It seems at the very least the third scenario should be fixed so that users can catch MapperException, bringing it in line with the other two.  Beyond that, I'm not sure we want to make strict guarantees about what exception a user does or does not find when they call `getCause()` In concrete terms that would greatly reduce the kind of changes we can make to the internals.

The most pragmatic perspective I can imagine is that we simply guarantee their exception will be in the `getCause()` chain and they can interate through it to find it, but should not expect it to be at a specific position.

Thoughts?

For visibility, I have till Friday, then it's a couple weeks before I can do more.  We may need to break this into two goals; tweaks if any to get this PR in; possible longer-term changes.

One very simple option could be we simply make the next release 1.3 instead of 1.2.19 and say, "it's different, deal with it"

Reply:
Was thinking the same about the versioning but the impact is that we should probably do it for all exceptions to not look half baked - this is why i proposed to stick to a backward compat option and refine the try/catch in nested calls and not around the whole call  but Im happy with the 1.3 one if coverage is reached too.

To refine my qtatement, issue is not having directly or not the exception but more about not rewrapping it in a minor version. Concretely if today i know my cause is the 3rd cause then it must not become the 4th in 1.2.19 - ok in 1.3. similarly for the type - cause due to jsonb abstraction, to stay portable, classname is sometimes checked.

Still thinking out loud but think we should just lean toward 1.3 here to ease things for everyones, wdyt?

Reply:
I think it's very fair to be concerned about people's exception handling in a patch release, no matter how brittle.  I'd probably lean toward the 1.3 as well so we had the flexibility to make changes we feel are the best changes.  What kind of coverage do we think we'd want for a 1.3?  Happy to hack up a serious number of tests :)  I just need a list of scenarios to work from.

Reply:
I'd say getters too even if here we should probably invent something or just say we accept tostring usage (think it is not good but can we do much better?). Can need an api review (basically each time user code is used it can happen so starting from the API enables to be exhaustive without trying to do it from the code - it is what I'd do at least).

Reply:
I'm not sure I understood the last sentence.  When you say through API review without trying to do it in code, what do you have in mind?

Reply:
Checking jsonb+mapper user api you get all user extensions points outside getters/setters/constructors (thinking to adapters, serializers etc).

Reply:
Cool.  I'll bring the Mapper API tests from PR #90 into this PR expand everything to cover the getters.  I have tests for the constructor for the bean, but now that we're talking I likely need to add tests for the constructor of the converter.  Is there a particular test or doc that would help me see how adapters and serializers work?

In these tests do we want to verify the exact position of the exception in the cause-chain or do we want to simply verify it is in the chain (i.e. not lost/swallowed)?

Reply:
If we target 1.3 I would let you choose, if we target 1.2.x we should ensure position are iso, and instances are user ones.

side note: maybe `@JsonbCreator` can be added for json-b too (factory method + constructor).

For mapper doc is there https://johnzon.apache.org/ (`@JohnzonConverter` but there are a few supported converter types - 3 IIRC), for jsonb it is the spec. If you need there should be test for all of these proprietary api.

Reply:
Let's target 1.3 and take the stance that users can guarantee their exception will be in the chain, but not necessarily where.  I'll write tests that ensure the exact user's exception instance is present in 100% of the covered cases and work to get coverage for literally everything I can find.

Adding @JsonbCreator to the list.  Is there a Mapper equivalent to watch out for?

Reply:
@dblevins not for factry methods, for constructor there was `ConstructorProperties` (to select one of the constructors if no no-arg one exist) and there is JsonbAccessMode programmatic hook (Factory) which enables any impl of a factory but no annotation.